### PR TITLE
Add RDP support [CWC-2453]

### PR DESCRIPTION
**Pull request by tflpd**
## Description of the change

This PR enables RDP connections to Windows agents utilizing the DB plugin which will soon be transformed to TCP plugin.

Also updates `TargetType` and `AgentType` so that we show the new "Linux" / "Windows" / "Kubernetes" targets to users

Left TODO: Add zli ref branch

## Testing

Add/edit a policy to have the RDP verb in it and try to zli connect to one of your Windows agents. Use the localhost:port returned by the zli in your RDP client.

**backend branch:** feat/rdp
**bzero branch:** develop
**charts branch:** 

#### Ready to run system tests?

- [x] Yes

## Relevant release note information

Release Notes: Add RDP support

## Related JIRA tickets

Relates to JIRA: CWC-2453

## Relevant HotFix if applicable

Relevant HotFix PR Number:

## Have you considered the security impacts?

Does this PR have any security impact?

- [ ] Yes
- [x] No

If yes, please explain:

Does this PR add a new/modifies the name/arguments of a `zli` command?

- [ ] Yes
- [ ] No

If yes, please provide the docs/zli-reference-manual branch reflecting the changes:

### CODE REVIEW

File: src/handlers/connect/connect.handler.ts [hunk 1]
```diff
@@ -53,8 +55,21 @@ export async function connectHandler(
 
         switch(createUniversalConnectionResponse.targetType)
         {
-        case TargetType.SsmTarget:
         case TargetType.Bzero:
+            if(createUniversalConnectionResponse.agentType == AgentType.Windows)
```
Comment (johncmerfeld) [inline src/handlers/connect/connect.handler.ts:59]: This probably warrants an explanatory comment, explaining that RDP is currently the only way to access windows targets, and in future iterations we'll require disambiguation of such a connection

File: src/handlers/disconnect/disconnect.handler.ts [hunk 2]
```diff
@@ -45,6 +45,11 @@ export async function disconnectHandler(
         const dbDaemonManagementService = newDbDaemonManagementService(configService);
         await handleDisconnect(dbDaemonManagementService, logger);
     }
+
+    if (targetType == 'all') {
```
Comment (johncmerfeld) [inline src/handlers/disconnect/disconnect.handler.ts:49]: Let's comment here explaining that `rdp` targets don't exist yet, so currently the only way to disconnect is by using `'all'` -- but they will exist soon!

File: src/handlers/list-daemons/list-daemons.handler.ts [hunk 3]
```diff
@@ -29,6 +29,10 @@ export async function listDaemonsHandler(
         await dbStatusHandler(configService, logger);
     }
 
+    if (targetType == 'all') {
```
Comment (johncmerfeld) [inline src/handlers/list-daemons/list-daemons.handler.ts:32]: Perhaps similar comment as in `disconnect` regarding the status of RDP targets

**Review by johncmerfeld — APPROVED** [review]
Need to do a bit more manual testing this afternoon, but this looks great. Just have a few small things.

File: src/handlers/connect/connect.handler.ts [hunk 4]
```diff
@@ -53,8 +55,21 @@ export async function connectHandler(
...
+            else {
+                const exitCode = await shellConnectHandler(TargetType.Linux, createUniversalConnectionResponse.targetUser, createUniversalConnectionResponse, configService, logger, loggerConfigService);
+
+                if (exitCode !== 0) {
+                    const errMsg = handleExitCode(exitCode, createUniversalConnectionResponse);
+                    if (errMsg.length > 0) {
+                        logger.error(errMsg);
+                    }
+                }
+                return exitCode;
```
[the remark this answers is not in the document]
Comment (ymarcus93) [reply src/handlers/connect/connect.handler.ts:75]: I noticed the same thing. This code is shared between a couple cases here. We could also move it to a separate function and call it, or do what you said above.

File: src/handlers/disconnect/disconnect.handler.ts (hunk 2 above)
Comment (ymarcus93) [reply src/handlers/disconnect/disconnect.handler.ts:49]: Can we also use `zli close` like the other connection types?

File: src/handlers/list-daemons/list-daemons.handler.ts [hunk 5]
```diff
@@ -29,6 +29,10 @@ export async function listDaemonsHandler(
...
+        await rdpStatusHandler(configService, logger);
+    }
```
Comment (ymarcus93) [inline src/handlers/list-daemons/list-daemons.handler.ts:34]: Wait I'm confused, why can't we add `targetType == rdp` in this PR? From what I understand, there is no RDP target yet, but `disconnect` is just used to kill daemons and we do have an RDP daemon and RDP daemon management service. So can we use that or no?

**Review by ymarcus93 — COMMENTED** [review]
Looks good. Couple of questions. Still testing on my end.

File: src/handlers/connect/connect.handler.ts (hunk 1 above)
Comment (tflpd) [reply src/handlers/connect/connect.handler.ts]: 16262d1e

File: src/handlers/connect/connect.handler.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (tflpd) [reply src/handlers/connect/connect.handler.ts]: Took a bit diff approach
89a8e0b8

File: src/handlers/connect/rdp-connect.handler.ts [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    };
+
+    const runtimeConfig = { ...baseEnv, ...pluginEnv, ...actionEnv };
+
+    let cwd = process.cwd();
+
+    // Copy over our executable to a temp file
+    let finalDaemonPath = '';
+    let args: string[] = [];
+    if (process.env.ZLI_CUSTOM_DAEMON_PATH) {
```
[the remark this answers is not in the document]
Comment (tflpd) [reply src/handlers/connect/rdp-connect.handler.ts]: faf57a1c

File: src/handlers/close-connection/close-connection.handler.ts [hunk 7]
```diff
@@ -29,8 +29,8 @@ export async function closeConnectionHandler(
                 await handleShell();
                 break;
             case 'db':
-                logger.info('Closing all db connections');
-                await closeConnections(configService, logger, 'db');
+                logger.info('Closing all rdp connections');
+                await closeConnections(configService, logger, 'rdp');
```
Comment (ymarcus93) [inline src/handlers/close-connection/close-connection.handler.ts:33]: This is for `db` not `rdp`. Was this change intentional?

File: src/handlers/disconnect/disconnect.handler.ts (hunk 2 above)
Comment (tflpd) [reply src/handlers/disconnect/disconnect.handler.ts:51]: 4bbc22bc

File: src/handlers/close-connection/close-connection.handler.ts (hunk 7 above)
Comment (tflpd) [reply src/handlers/close-connection/close-connection.handler.ts]: 4bbc22bc

File: src/handlers/list-daemons/list-daemons.handler.ts (hunk 3 above)
Comment (tflpd) [reply src/handlers/list-daemons/list-daemons.handler.ts:34]: 1bf3429f

File: src/handlers/attach/attach.handler.ts [hunk 8]
```diff
@@ -42,7 +42,7 @@ export async function attachHandler(
 
     if(connectionSummary.targetType == TargetType.SsmTarget || connectionSummary.targetType == TargetType.DynamicAccessConfig) {
         return createAndRunShell(configService, logger, connectionSummary);
-    } else if(connectionSummary.targetType == TargetType.Bzero) {
+    } else if(connectionSummary.targetType == TargetType.Linux) {
```
[the remark this answers is not in the document]
Comment (tflpd) [reply src/handlers/attach/attach.handler.ts:45]: @johncmerfeld

File: src/handlers/close-connection/close-connection.handler.ts [hunk 9]
```diff
@@ -48,6 +52,7 @@ export async function closeConnectionHandler(
             const results = await Promise.allSettled([
```
Comment (ymarcus93) [inline src/handlers/close-connection/close-connection.handler.ts:52]: Can we update this log statement to include `rdp`? https://github.com/bastionzero/zli/blob/4bbc22bc01189a14766683b719948255a54657c5/src/handlers/close-connection/close-connection.handler.ts#L50

File: src/handlers/list-daemons/list-daemons.handler.ts (hunk 5 above)
Comment (tflpd) [reply src/handlers/list-daemons/list-daemons.handler.ts:36]: I believe the answer is because MIke does not want to add an RDP target yet but I will cc @johncmerfeld to this

File: src/handlers/connect/connect.handler.ts [hunk 10]
```diff
@@ -13,6 +13,10 @@ import { startKubeDaemonHandler } from 'handlers/connect/kube-connect.handler';
 import { MixpanelService } from 'services/Tracking/mixpanel.service';
 import { CreateUniversalConnectionResponse } from 'webshell-common-ts/http/v2/connection/responses/create-universal-connection.response';
 import { handleExitCode } from 'utils/daemon-utils';
+import { rdpConnectHandler } from './rdp-connect.handler';
+import { AgentType } from 'webshell-common-ts/http/v2/target/types/agent.types';
+
+const IGNORE_TARGET_USER_MSG: string = 'Specifying a target user or role for RDP, Web, or Database targets is not supported at this time. Ignoring target name / role and requesting connection per policy.';
```
Comment (ymarcus93) [inline src/handlers/connect/connect.handler.ts:19]: Is this message correct? If so, I don't see it being logged in `case TargetType.Db` below.

File: src/handlers/disconnect/disconnect.handler.ts (hunk 2 above)
Comment (ymarcus93) [reply src/handlers/disconnect/disconnect.handler.ts:51]: Close support looks good

File: src/handlers/connect/connect.handler.ts (hunk 1 above)
Comment (ymarcus93) [reply src/handlers/connect/connect.handler.ts]: Thanks! Going to bookmark this so I don't forget

File: src/handlers/attach/attach.handler.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (johncmerfeld) [reply src/handlers/attach/attach.handler.ts:45]: @ymarcus93  you're right, this should be `Bzero` because that's what the connection type will return. I'll push a fix

However, due to CWC-1826, this is actually moot because we don't support attaching to bzero shells anyway

File: src/handlers/list-daemons/list-daemons.handler.ts (hunk 5 above)
Comment (johncmerfeld) [reply src/handlers/list-daemons/list-daemons.handler.ts:33]: 8374986

File: src/handlers/disconnect/disconnect.command-builder.ts [hunk 11]
```diff
@@ -5,12 +5,12 @@ export type disconnectArgs = {targetType: string};
...
+            choices: ['kube', 'db', 'web', 'rdp', 'all'],
             nargs: 1,
             type: 'string',
             default: 'all',
             requiresArg: false,
         })
-        .example('$0 disconnect', 'Disconnect all local Zli Daemon')
-        .example('$0 disconnect kube', 'Disconnect Kube local Zli Daemon');
+        .example('$0 disconnect', 'Disconnect all open connections')
+        .example('$0 disconnect kube', 'Disconnect all Kubeneretes connections');
```
Comment (ymarcus93) [inline src/handlers/disconnect/disconnect.command-builder.ts:15]: This should still say daemons to differentiate between `zli close` vs. `zli disconnect`. `zli close` closes open connections made at any machine where the user is logged in. `zli disconnect` is based on local machine, hence the use of `daemon` in the previous example string.

I'm also fine with a different qualifier, but I think the distinction should still exist.

File: src/handlers/list-daemons/list-daemons.handler.ts [hunk 12]
```diff
@@ -28,10 +28,7 @@ export async function listDaemonsHandler(
     if (targetType == 'all' || targetType == 'db') {
         await dbStatusHandler(configService, logger);
     }
-
-    // As we are transitioning to our new Targets notion, RDP targets do not exist yet, thus the only way
-    // to close such connection is by closing all
-    if (targetType == 'all') {
+    if (targetType == 'all' || targetType == 'all') {
```
Comment (ymarcus93) [inline src/handlers/list-daemons/list-daemons.handler.ts:31]: I think you meant `|| targetType == 'rdp'`

**Comment by johncmerfeld** [comment]
> zli policy create-tconnect -n windows-policy -u yuval@bastionzero.com -v rdp

Good catch! Nope, for me I see: `Missing required argument: targetUsers`

**Comment by ymarcus93** [comment]
> > zli policy create-tconnect -n windows-policy -u [yuval@bastionzero.com](mailto:yuval@bastionzero.com) -v rdp
> 
> Good catch! Nope, for me I see: `Missing required argument: targetUsers`

Oh whoops I forgot some flags in my example. But this also doesn't work for me:

`zli policy create-tconnect -n windows-policy-2 -u yuval@bastionzero.com -v rdp --targetUsers root -e Default`

It creates the policy with shell verb but not RDP.

**Comment by ymarcus93** [comment]
Is `targetUser` required if I only have an RDP policy? I think it used to be but on webapp it lets me create the policy without specifying target user. Doesn't seem to be the case in the `zli` given that error message you got

**Comment by johncmerfeld** [comment]
> Is `targetUser` required if I only have an RDP verb? I think it used to be but on webapp it lets me create the policy without specifying target user. Doesn't seem to be the case in the `zli` given that error message you got

The correct behavior would be to allow the policy to be created without a target user, so what I saw was incorrect. Looks like two different errors between us.

File: src/handlers/list-daemons/list-daemons.handler.ts (hunk 12 above)
Comment (johncmerfeld) [reply src/handlers/list-daemons/list-daemons.handler.ts:31]: Whoops, thought I fixed this!

File: src/handlers/disconnect/disconnect.command-builder.ts (hunk 11 above)
Comment (johncmerfeld) [reply src/handlers/disconnect/disconnect.command-builder.ts:15]: Ah, fair enough, thanks for keeping me honest

Comment (ymarcus93) [reply src/handlers/disconnect/disconnect.command-builder.ts]: Your change though did reveal something though. The original language is not good grammar, probably remnant before we had multi-daemon connections on single machine.

It should say `disconnect all local zli daemons` and `disconnect all Kube local zli daemons` respectively.

File: src/handlers/connect/connect.handler.ts (hunk 10 above)
Comment (johncmerfeld) [reply src/handlers/connect/connect.handler.ts:19]: It's not logged for databases because in that case, you can specify a target user if you're using a SplitCert connection. @tflpd might be worth confirming this phrasing with Ann if you haven't already. It's a nuanced situation

Comment (tflpd) [reply src/handlers/connect/connect.handler.ts:19]: Yuval's point is that this mentions DBs but is not used in DBs case I believe

Comment (tflpd) [reply src/handlers/connect/connect.handler.ts]: 5dc26365..709d9aee

**Review by ymarcus93 — APPROVED** [review]
LGTM. Tested on my Mac
